components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -1122,240 +1332,125 @@ void FeedV2Builder::NotifyUpdateCompleted() {
-      if (TossCoin()) {
-        DVLOG(1) << "Step 5: Standard Block";
-        items = GenerateBlockFromContentGroups(articles, locale, publishers,
-                                               eligible_content_groups);
-      } else {
-        items = GenerateClusterBlock(locale, publishers, subscribed_channels,
-                                     topics, articles);
-      }
-    } else if (iteration_type == 2) {
-      // Step 6: Optional special card
-      // https://docs.google.com/document/d/1bSVHunwmcHwyQTpa3ab4KRbGbgNQ3ym_GHvONnrBypg/edit#heading=h.n1ipt86esc34
-      if (TossCoin()) {
-        DVLOG(1) << "Step 6: Special Block";
-        items = GenerateSpecialBlock(suggested_publisher_ids);
-      } else {
-        DVLOG(1) << "Step 6: None (approximately half the time)";
-      }
-    } else {
-      NOTREACHED();
-    }
+void FeedV2Builder::GenerateFeed(UpdateSettings settings,
+                                 mojom::FeedV2TypePtr type,
+                                 FeedGenerator generator,
+                                 BuildFeedCallback callback) {
+  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
+  UpdateData(
+      std::move(settings),
+      base::BindOnce(
+          [](const base::WeakPtr<FeedV2Builder> builder,
+             mojom::FeedV2TypePtr type, FeedGenerator generate,
+             BuildFeedCallback callback, base::OnceClosure on_complete) {
+            auto built_callback = base::BindOnce(
+                [](BuildFeedCallback callback,
+                   base::OnceClosure completion_callback,
+                   mojom::FeedV2Ptr feed) {
+                  std::move(callback).Run(std::move(feed));
+                  std::move(completion_callback).Run();
+                },
+                std::move(callback), std::move(on_complete));
 
-    // If we couldn't generate a normal block, break.
-    if (iteration_type == 0 && items.empty()) {
-      break;
-    }
+            if (!builder) {
+              std::move(built_callback).Run(mojom::FeedV2::New());
+              return;
+            }
 
-    DVLOG(1) << "Adding " << items.size() << " new articles (iteration type is "
-             << iteration_type << "). Currently have " << feed->items.size()
-             << " articles";
-    add_items(items);
-    ++iteration;
-  }
+            const auto& publishers =
+                builder->publishers_controller_->GetLastPublishers();
+            const auto& locale =
+                builder->publishers_controller_->GetLastLocale();
+            std::vector<std::string> channels;
+            for (const auto& [channel_id, channel] :
+                 builder->channels_controller_->GetChannelsFromPublishers(
+                     publishers, &*builder->prefs_)) {
+              if (base::Contains(channel->subscribed_locales, locale)) {
+                channels.push_back(channel_id);
+              }
+            }
 
-  return feed;
+            FeedGenerationInfo info(
```

Comment:
I wonder what it's the typical size of `FeedGenerationInfo` and how much CPU we need to construct it (including all the for-loops here.

Reply:
I imagine the typical size is large. 3500 `FeedItems`, similar number of `TopicArticles`, and maybe ~250 channels/publishers.

That said, I just timed this on my machine and it takes 11ms for this bit of code, which constructs the `FeedGenerationInfo`:
![image](https://github.com/brave/brave-core/assets/7678024/a848bdb4-71e3-4f0f-8145-391aebab212b)

Reply:
Sadly, 11ms also sounds like an issue for UI thread, especially for Android. UI thread is in charge of making frames each 33ms on average, therefore heavy computation is prohibited here.
From [the upstream doc](https://source.chromium.org/chromium/chromium/src/+/main:docs/threading_and_tasks.md): 
```
Do not perform expensive computation or blocking IO on the main thread
(a.k.a. “UI” thread in the browser process) or IO thread (each process's thread for receiving IPC).
A busy UI / IO thread can cause user-visible latency, so prefer running that work on the thread pool.
```

I believe the only way to solve this issue completely is to move all the inner things to a dedicated sequence (aka virtual thread in the pool).
[A good example](https://source.chromium.org/chromium/chromium/src/+/main:components/subresource_filter/content/browser/async_document_subresource_filter.h;l=188-189?q=async_document_subresource_filter.h) in Chromium code.
Otherwise, it's about impossible to manipulate such large structures on UI thread.

I'm not against solving the issue step-by-step, but I'd like to discuss and sync the final plan before we going to merge an individual PRs

Reply:
Yeah agreed, we should fix the root cause (that we're on the UI thread) rather than just putting bandaids over things. However, I think it's going to be a much more significant refactor to move everything into a sequence.

What I'm roughly thinking at the moment is this:
1. We land & uplift this PR, roughly as is.
2. We refactor `BraveNewsController` to have it's own sequence. Every call to subservices should happen on this sequence (this is going to be complicated because things are quite tightly tied to `PrefService` at the moment).
3. We change this `FeedGenerationInfo` to just take a reference (still handy to tie everything together) and get rid of the `base::ThreadPool` usages here.

Does that sound reasonable? Happy to jump on a call to discuss :smile:

Reply:
Okay, after a bit more research, I think something a bit like what `PasswordStore` does is probably the way to go.

We introduce a `BraveNewsControllerFrontEnd`. Lives on the UI thread, has a `scoped_refptr` to a `BraveNewsController` and `SequencedTaskRunner` which is posts tasks to the `BraveNewsController` on.

Reply:
Let's merge this PR (after we sure that all the callback are good), this should improve the situation.
The next steps are to solve the issue completely. I hope I could provide you some perf metrics to track the progress (in https://github.com/brave/brave-core/pull/21876)